feat(avm)!: Calldata Hashing

MirandaWood opened the pull request:
## Constraining Calldata Hashing

For a detailed overview + diagrams look at the design document! Here is a brief-ish overview. Following a similar path to bytecode hashing (#16756) , we use the `poseidon2` trace to prove correctness of calldata hash and the calldata size.

When we initialise a `context` with a new public call request, we emit a `CalldataEvent`:

```C++
struct CalldataEvent {
    uint32_t context_id;
    uint32_t calldata_size;
    std::vector<FF> calldata;
    // FF output_hash; // NOTE - removed as it was unused, should we emit this as well as a hint?
};
```

This data makes its way to `calldata.pil` for retrieval (from which `data_copy.pil` extracts the fields) and `calldata_hashing.pil` for hashing and size constraining. Each field in `calldata` is looked up and matched to a field in `poseidon2`. At the final (`latch`) row, we constrain that all calldata fields have been included (no more, no less), the `output_hash` matches the `output` of the posiedon hash, and the given `calldata_size` is correct.

This PR adds a new lookup from `tx.pil` to `calldata_hashing.pil` linking the hash and size from the corresponding public call request. Now, the existing lookup to the **hash** in the public inputs is fully constrained to be correct. 

It also adds `calldata_size` as a new column (and a new property in the `EnqueuedCallEvent`) and appends it to the existing lookup to `execution.pil` at dispatch, constraining that `execution.parent_calldata_size` (taken from `context.pil`) matches. This value is also looked up by `data_copy`, and so is now fully constrained.

### Main TODOs/Open Qs:

- [ ] ~Check e2e tests~ + create test with explicit multiple calldata reads
- [ ] Finalise approach of `read_addr_plus_one`
- [x] Finalise removal of hash from calldata event
- [ ] How best to combine some relations in `calldata_hashing` - i.e. `#[START_AFTER_LATCH]` and `    #[START_INDEX_IS_ZERO]`? `#[INDEX_INCREMENTS_i]`?

MirandaWood commented on barretenberg/cpp/pil/vm2/data_copy.pil:
```diff
@@ -307,14 +307,16 @@ namespace data_copy;
+
     #[COL_READ]
-    cd_copy_col_read { value, dst_context_id, read_addr }
+    cd_copy_col_read { value, dst_context_id, read_addr_plus_one }
```
I thought this was the lowest footprint way of looking up the correct values in the `calldata` trace (which starts at `i = 1`) without messing with memory addresses, but it is a bit ugly. Any thoughts?

jeanmon replied:
Alternatively, the index in calldata.pil should start with zero but I guess that it is very annoying to change. I think it is fine. Just add a comment about the fact that we start indexing by one in calldata.

MirandaWood replied:
Commented just above the def of `read_addr_plus_one`, will add the alias todo!

jeanmon commented on barretenberg/cpp/pil/vm2/calldata.pil:
```diff
@@ -26,5 +31,5 @@ namespace calldata;
     #[CONTEXT_ID_CONTINUITY]
     (1 - precomputed.first_row) * sel * (1 - latch) * (context_id - context_id') = 0;
-    
+
```
@MirandaWood Do you know the answer to this question? Maybe we can delete the TODO

jeanmon replied:
I think there is a potential issue that one could add calldata entries which are not used for hashing but pass to data_copy.pil.

I would order by increasing order of `context_id` (probably already the case in tracegen) as follows:
pol commit diff_context_id = latch * sel' * (context_id' - context_id);
`latch { diff_context_id }  in precomputed.sel_range_16 { precomputed.clk }`

Note that context_id's are assumed to be 16 bits. (We used this to constrain memory).

MirandaWood replied:
Short answer: no, I don't know!

> I think there is a potential issue that one could add calldata entries which are not used for hashing but pass to data_copy.pil.

I think that may be possible, but we'd have to bypass the usual tx execution path and only use the calldata event in the `calldata.pil` trace - I feel like this would fail somewhere around dispatch, then copy couldn't run, but I'm unsure! Better to be safe anyway.

Happy to add this but I don't think tracegen sorts at the moment, so I will implement that as well. Would checking the diff is non-0 using the usual inverse method be preferable to a range check?

jeanmon replied:
> I feel like this would fail somewhere around dispatch
 
 This is assuming a malicious prover (soundness) and therefore you cannot assume our code path with dispatch, etc, .... The tracegen can be completely different than our "honest tracegen".
 
>  Would checking the diff is non-0 using the usual inverse method be preferable to a range check?
 
 This would not be safe because you can overflow the field and go back to the same context_id.
 
 By requiring a "small increment" bounded by 16-bit, we know that we cannot reach p which is so much bigger.

jeanmon replied:
Note also a mistake on my side. We must ensure at `latch` that we increment, so, we have to compute the diff minus one:

pol commit diff_context_id = latch * sel' * (context_id' - context_id - 1);

jeanmon commented on barretenberg/cpp/pil/vm2/calldata.pil:
```diff
@@ -26,5 +31,5 @@ namespace calldata;
     // Context id does not change until we latch
     #[CONTEXT_ID_CONTINUITY]
     (1 - precomputed.first_row) * sel * (1 - latch) * (context_id - context_id') = 0;
-    
```
We can remove the multiplicative factors `(1 - precomputed.first_row) * sel` in #[CONTEXT_ID_CONTINUITY] and replace `(1 - latch)` by `(1 - FIRST_OR_LAST_CALLDATA)`

MirandaWood replied:
Oops yep - some old code I didn't look over properly, thanks!

jeanmon commented on barretenberg/cpp/pil/vm2/calldata_hashing.pil:
```diff
@@ -12,24 +14,180 @@ namespace cd_hashing;
 
     pol commit context_id;
-    pol commit length_remaining;
+    #[ID_CONSISTENCY]
```
Please move #[ID_CONSISTENCY] and #[SIZE_CONSISTENCY] after the definition of `latch` and `LATCH_CONDITION`

jeanmon commented on barretenberg/cpp/pil/vm2/calldata.pil:
```diff
@@ -1,7 +1,6 @@
 include "precomputed.pil";
 include "calldata_hashing.pil";
 
-// Calldata is largely going to be constrained by calldata hashing
```
Adding a usage of this trace (lookup to) and a little example table would be great.

jeanmon commented on barretenberg/cpp/pil/vm2/calldata_hashing.pil:
```diff
@@ -1,8 +1,10 @@
 include "poseidon2_hash.pil";
+include "constants_gen.pil";
+
+namespace calldata_hashing;
```
Same here, adding a usage and little table example would help.

jeanmon commented on barretenberg/cpp/pil/vm2/calldata_hashing.pil:
```diff
@@ -12,24 +14,180 @@ namespace cd_hashing;
+
+    // Needs to be a committed column as it is used in the lookup:
+    pol commit sel_not_start;
+    sel_not_start - sel * (1 - start) = 0;
```
```suggestion
    sel_not_start = sel * (1 - start) ;
```

jeanmon commented on barretenberg/cpp/pil/vm2/calldata_hashing.pil:
```diff
@@ -12,24 +14,180 @@ namespace cd_hashing;
+
+    // Each index in the array increments (note: ideally we would just use a single index, but we need to use
+    // index + 1 and index + 2 in the lookups to calldata.pil):
+    // TODO(MW): Possibly a cleverer way to combine some relations here
```
You could batch if there is an intrisic upper bound for index values. Assuming it is 2^31, you can batch with:

sel * (index[1] - (index[0] + 1) + 2^32 * ( index[2] - (index[1] + 1) )) = 0;

Here it is clearly not worth it I would say. We are not saving much.

jeanmon commented on barretenberg/cpp/pil/vm2/calldata_hashing.pil:
```diff
@@ -12,24 +14,180 @@ namespace cd_hashing;
+    pol commit index[3];
+
+    // We must start at index == 0:
+    // TODO(MW): Since we have constrained that a START must exist after a latch, can we combine this w another relation?
```
You can modify #[START_AFTER_LATCH] to be:

`sel' * (start' - LATCH_CONDITION + 2 * index[0]') = 0;`

but we need to have an intrinsic upper bound on index[0] such as 2^128.
Furthermore, this adds a shifted column for index[0] so I am skeptical it is worth it.
I would not do it.

MirandaWood replied:
I considered something like this but yeah - wanted to avoid anything bounded. This comment more refers to combining relations like `START_INDEX_IS_ZERO` and `START_IS_SEPARATOR` given that a `start` row must exist. I think anything would just add more shifted columns and/or checks so not worth it?

jeanmon replied:
It is very readable and these ones are so simple that the gain would be negligible to try to combine them. I would keep it like this.

jeanmon commented on barretenberg/cpp/pil/vm2/calldata_hashing.pil:
```diff
@@ -12,24 +14,180 @@ namespace cd_hashing;
+    #[PADDING_END]
+    PADDING_2 * (1 - latch) = 0;
+
+    // padding_1 == 1 & padding_2 == 1 <==> calldata_size = index[0]
```
We can add a comment that these conditions are valid only on row with `latch == 1`

MirandaWood replied:
Tbh, I thought `PADDING_END` and the use of `FINAL` (plus comment on line 96) was sufficient - I can move the line 96 comment further down?

jeanmon replied:
My point is that I have seen some footgun where relations are conditioned by another condition (e.g. latch == 1) and the this assumption is forgotten and we assume that it is true on other rows.
However, I am realizing that here you do not need the conditioning except in #[CHECK_FINAL_INDEX].
So the risk is more completeness than soundness. Maybe just add a comment that all the above relations are fine even on non-latch row because there will be no padding and PADDING_1 == 0 and PADDING_2 == 0 as per our trace generation.

jeanmon commented on barretenberg/cpp/pil/vm2/calldata_hashing.pil:
```diff
@@ -12,24 +14,180 @@ namespace cd_hashing;
+    // We use it to ensure the ordering of poseidon_2 rounds is correct (i.e. a malicious prover cannot swap the order of poseidon rounds).
+    pol commit rounds_rem;
+
+    pol PADDING_DIFF = 3 * rounds_rem + index[0] - input_len;
```
I think it is correct because we can reasonably assume that rounds_rem is smaller than p/3 but strictly speaking, there could higher values for rounds_rem which can satisfy the next condition. Such values would be gigantic ( > p/3) and the lookup to poseidon would fail.

However, I would rather lean on having simple "decrement evolution" constraints like:

#[ROUNDS_REM_INIT]
latch * (rounds_rem - 1) = 0;

#[ROUNDS_REM_DECREMENT]
sel * (1 - LATCH_CONDITION) * (rounds_rem' - rounds_rem - 1) = 0;

MirandaWood replied:
Fair enough, I will switch to decrementing! Thanks :)

jeanmon commented on barretenberg/cpp/pil/vm2/data_copy.pil:
```diff
@@ -307,14 +307,16 @@ namespace data_copy;
     #[CD_COPY_COLUMN]
     cd_copy_col_read = SEL_PERFORM_COPY * (1 - padding) * is_top_level * sel_cd_copy;
 
+    // The calldata trace starts at index = 1 (TODO(MW): cleanup?):
```
Add a TODO about a migration to an alias once lookup supports them.

jeanmon commented on barretenberg/cpp/pil/vm2/tx.pil:
```diff
@@ -180,8 +181,20 @@ namespace tx;
+        calldata_hash,
+        calldata_size,
+        next_context_id
+    } in calldata_hashing.latch {
```
I think it is secure to have a lookup but it would be easier to reason about the soundness if it would be a permutation (unless we cannot have a permutation like when we run out of gas, etc, ...). If it is too tedious to be sure about this at this time, we can keep it like this.

MirandaWood replied:
Makes sense! I'll think about this, and at the least add a todo, thank you!

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/constraining/relations/calldata_hashing.test.cpp:
```diff
@@ -0,0 +1,768 @@
+                              "SIZE_CONSISTENCY");
+}
+
+} // namespace
```
Amazing work!
I did not see any test to cover the use case where there is no calldata (size == 0). Maybe there is no need.
How is this handled in the code?

MirandaWood replied:
I added one in simulation since in this case (according to some existing comments) we do not emit an event at all. IIRC I asked about this for bytecode hashing and essentially the circuit is not meant to handle empty bytecode, so it was left as is. I can see what happens with an input of size == 0, and ensure we cannot form an attack?

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/tracegen/calldata_trace.cpp:
```diff
@@ -45,39 +45,49 @@ void CalldataTraceBuilder::process_hashing(
-            input_size -= chunk_size;
-
+        FF output_hash = Poseidon2::hash(calldata_with_sep);
+        auto padding_amount = (3 - (calldata_with_sep.size() % 3)) % 3;
```
A bit hacky but `2 * (calldata_with_sep.size() % 3)` would do the job.

jeanmon replied:
If you were to implement this, it might require a small comment like we need this modulo 3 which corresponds to multiply by 2 modulo 3:
0 -> 0
1 -> 2
2 -> 1

I made a mistake above with the parenthesis, we need:

(2 * calldata_with_sep.size()) % 3

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/tracegen/calldata_trace.cpp:
```diff
@@ -45,39 +45,49 @@ void CalldataTraceBuilder::process_hashing(
+        auto padding_amount = (3 - (calldata_with_sep.size() % 3)) % 3;
+        auto num_rounds = (calldata_with_sep.size() + padding_amount) / 3;
+        uint32_t index = 0;
+        while (num_rounds > 0) {
```
Maybe rename `num_rounds` to `num_rounds_rem`

jeanmon commented on noir-projects/aztec-nr/aztec/src/hash.nr:
```diff
@@ -166,6 +166,19 @@ unconstrained fn compute_var_args_hash() {
+    dep::std::println(hash);
+    let hash_check = hash_calldata(input.as_slice());
+    assert(hash == hash_check);
+    assert(hash == 0x191383c9f8964afd3ea8879a03b7dda65d6724773966d18dcf80e452736fc1f3);
```
You could add a comment that you use this constant also in some .cpp tests and that we need to keep this one.

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/tracegen/tx_trace.cpp:
```diff
@@ -661,6 +662,7 @@ const InteractionDefinition TxTraceBuilder::interactions =
         .add<lookup_tx_read_phase_table_settings, InteractionType::LookupGeneric>()
         .add<lookup_tx_phase_jump_on_revert_settings, InteractionType::LookupGeneric>()
         .add<lookup_tx_read_phase_length_settings, InteractionType::LookupGeneric>()
+        .add<lookup_tx_read_calldata_hash_settings, InteractionType::LookupGeneric>()
```
The calldata_events are non deduplicated, could not we use the LookupSequential variant here?

MirandaWood replied:
Good point, I just copy pasted from the others here!

jeanmon reviewed (changes requested):
Fantastic work overall!

jeanmon commented on barretenberg/cpp/pil/vm2/calldata_hashing.pil:
```diff
@@ -3,6 +3,30 @@ include "precomputed.pil";
+//   index_0_ |  index_1_ | index_2_ | context_id | input_0_ | input_1_ | input_2_ | output_hash | latch | start | rounds_rem | padding | calldata_size |
+// -----------+-----------+----------+------------+----------+----------+----------+-------------+-------+-------+------------+---------+---------------+
+//     0      |     1     |     2    |      1     |    sep   |   0x111  |   0x222  |  cd_hash_0  |   0   |   1   |     2      |    0    |       4       |
+//     3      |     4     |     5    |      1     |   0x444  |     0    |     0    |  cd_hash_0  |   1   |   0   |     1      |    2    |       4       |
```
0x333 should be in input_0 and 0x44 at input_1_, no?

jeanmon replied:
And then padding will be 1

MirandaWood replied:
Oops yep, thanks!

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/tracegen/calldata_trace.cpp:
```diff
@@ -13,20 +14,37 @@ void CalldataTraceBuilder::process_retrieval(
+                          { C::calldata_latch, is_latch ? 1 : 0 },
+                          // Note that the diff is shifted by 1 to ensure the context_ids are increasing:
+                          { C::calldata_diff_context_id,
+                            is_latch && !is_last ? sorted_events[j + 1]->context_id - context_id - 1 : 0 },
```
MAybe add parenthesis around `is_latch && !is_last` ?

jeanmon reviewed (approved):
Very good work! I just have nitpick comments.

MirandaWood committed:
feat(avm)!: Calldata Hashing

## Constraining Calldata Hashing

For a detailed overview + diagrams look at the design document! Here is a brief-ish overview. Following a similar path to bytecode hashing (#16756) , we use the `poseidon2` trace to prove correctness of calldata hash and the calldata size.

When we initialise a `context` with a new public call request, we emit a `CalldataEvent`:

```C++
struct CalldataEvent {
    uint32_t context_id;
    uint32_t calldata_size;
    std::vector<FF> calldata;
    // FF output_hash; // NOTE - removed as it was unused, should we emit this as well as a hint?
};
```

This data makes its way to `calldata.pil` for retrieval (from which `data_copy.pil` extracts the fields) and `calldata_hashing.pil` for hashing and size constraining. Each field in `calldata` is looked up and matched to a field in `poseidon2`. At the final (`latch`) row, we constrain that all calldata fields have been included (no more, no less), the `output_hash` matches the `output` of the posiedon hash, and the given `calldata_size` is correct.

This PR adds a new lookup from `tx.pil` to `calldata_hashing.pil` linking the hash and size from the corresponding public call request. Now, the existing lookup to the **hash** in the public inputs is fully constrained to be correct.

It also adds `calldata_size` as a new column (and a new property in the `EnqueuedCallEvent`) and appends it to the existing lookup to `execution.pil` at dispatch, constraining that `execution.parent_calldata_size` (taken from `context.pil`) matches. This value is also looked up by `data_copy`, and so is now fully constrained.

### Main TODOs/Open Qs:

- [ ] ~Check e2e tests~ + create test with explicit multiple calldata reads
- [ ] Finalise approach of `read_addr_plus_one`
- [x] Finalise removal of hash from calldata event
- [ ] How best to combine some relations in `calldata_hashing` - i.e. `#[START_AFTER_LATCH]` and `    #[START_INDEX_IS_ZERO]`? `#[INDEX_INCREMENTS_i]`?
